packages/wrangler/src/deploy/deploy.ts

```diff
@@ -965,6 +1001,88 @@ function deployWfpUserWorker(
 	logger.log("Current Deployment ID:", deploymentId);
 }
 
+interface SourceMapV3 {
+	version: 3;
+	file: string;
+	sourceRoot: string;
+	sources: string[];
+	sourcesContent: string[];
+	mappings: string;
+	names: string[];
+}
+
+/**
+ * Load and normalize a source map emitted by Wrangler using the given path and
+ * directory metadata.
+ */
+function loadBundledSourceMap(
+	{ name, filePath }: CfModule,
+	sourceMapPath: string,
+	metadata: SourceMapMetadata
+): CfWorkerSourceMap[] {
+	if (filePath === undefined) {
+		return [];
+	}
+	const map = JSON.parse(readFileSync(sourceMapPath).toString()) as SourceMapV3;
+	// Overwrite the file property of the source map to match the name of the
+	// main module in the multipart upload.
+	map.file = name;
+	// Remove the temporary directory prefix generated by Wrangler that appears
+	// in the source root path.
+	map.sourceRoot = map.sourceRoot.slice(metadata.tmpDir.length);
+	map.sources = map.sources.map((source) => {
+		const originalPath = path.join(path.dirname(filePath), source);
+		// Remove the temporary build directory prefix generated by Wrangler
+		// that appears in the source path.
+		return originalPath.slice(metadata.entryDirectory.length + 1);
+	});
+	return [
+		{
+			name: name + ".map",
+			content: JSON.stringify(map),
+		},
+	];
+}
+
+/**
+ * Find source maps by scanning module contents for special `//#
+ * sourceMappingURL=` comments pointing to source map files.
+ */
+function scanSourceMaps(modules: CfModule[]): CfWorkerSourceMap[] {
+	const maps: CfWorkerSourceMap[] = [];
+	for (const module of modules) {
+		const commentPrefix = "//# sourceMappingURL=";
+		const commentIndex = module.content.indexOf(commentPrefix);
+		if (commentIndex === -1 || module.filePath === undefined) {
+			continue;
+		}
+		// Assume the source map path in the comment is relative to the
+		// generated file it appears in.
+		const commentPath = module.content
+			.slice(commentIndex + commentPrefix.length)
+			.toString()
+			.trim();
+		// Convert source map path to an absolute path that we can read.
+		const wranglerPath = path.join(path.dirname(module.filePath), commentPath);
+		if (!existsSync(wranglerPath)) {
+			throw new Error(
+				`Invalid source map path in ${module.filePath}: ${wranglerPath} does not exist.`
+			);
+		}
+		const map = JSON.parse(
+			readFileSync(wranglerPath).toString()
+		) as SourceMapV3;
+		// Overwrite the file property of the sourcemap to match the name of the
+		// corresponding module in the multipart upload.
+		map.file = module.name;
```

Comment:
Do we want to update the `sources` here too? When testing with a custom build command with esbuild (`pnpm esbuild --outdir=dist --sourcemap=external src/**/*.ts`) I ended up with entries in my sources that contained one or more `../`'s with paths like `../../src/foo/bar.ts`. We want to display these as `./src/foo/bar.ts` *one day* in the quick editor so we probably want to do that transformation here in wrangler. I guess a naive call to `substring` if the string starts with `../` would suffice?

Reply:
According to the spec I think that the plan is only to support inline content for the sources?

Reply:
@zebp Pushed a change to clean up the `sources` array as you suggested. The custom build part of this PR still feels pretty brittle, and I'm not sure how much we can do about it.

@petebacondarwin The plan was to only to support ["linked"](https://esbuild.github.io/api/#sourcemap) source maps. We could support inline as well but I think we want to deter people from unnecessarily increasing their JS bundle size.